### PR TITLE
Extend AM validate_length_of matcher to support array attributes

### PR DESCRIPTION
**Pull request by jarenas9539**
First step to close: https://github.com/thoughtbot/shoulda-matchers/issues/1559

# Solution

Were added two mechanisms to treat `validate_length_of_matcher` related attribute as array. One was through `array_column?` helper(also used on `validate_absence_of_matcher`) based on column type if `array?` helper is defined on model's `columns_hash` and another one with an explicit chain method `as_array` that will force to validate column with arrays values.
Why this way? First one will cover pure `array: true` columns for postgresql, second one could be dynamically used to check, for example, json columns storing arrays and being treated like that. This second approach will leave more freedom when using the validator.

Main validate_length_of_matcher specs were duplicated for the two variants. After all review's iteration for this PR concluded and if it's finally approved and merged this approach, I'll continue extending the array support to other validators(as inclusion/exclusion).

### CODE REVIEW

File: lib/shoulda/matchers/active_model/validate_absence_of_matcher.rb [hunk 1]
```diff
@@ -143,12 +143,6 @@ def reflection
             @subject.class.reflect_on_association(@attribute)
         end
 
-        def array_column?
```
Comment (mcmire) [inline lib/shoulda/matchers/active_model/validate_absence_of_matcher.rb:146]: Since this PR is focused on updating the `validate_length_of` matcher, should we do this in a different PR? This will just override the method you've added to ValidationMatcher, but that's okay for now.

File: lib/shoulda/matchers/active_model/validate_length_of_matcher.rb [hunk 2]
```diff
@@ -451,15 +456,19 @@ def allow_nil_does_not_match?
...
-          allows_value_of(string_of_length(length), message)
+          allows_value_of(element_of_length(length), message)
         end
 
         def disallows_length_of?(length, message)
-          disallows_value_of(string_of_length(length), message)
+          disallows_value_of(element_of_length(length), message)
+        end
+
+        def element_of_length(length)
```
Comment (mcmire) [inline lib/shoulda/matchers/active_model/validate_length_of_matcher.rb:466]: "Element" in this context sounds like it would be an element in an array, which isn't quite accurate. What do you think about `value_of_length`?

File: spec/unit/shoulda/matchers/active_model/validate_length_of_matcher_spec.rb [hunk 3]
```diff
@@ -370,15 +370,205 @@ def configure_validation_matcher(matcher)
...
+        expect(validating_length(type: :jsonb, is: 4)).
+          not_to validate_length_of(:attr).as_array.is_equal_to(5).with_message(/.*/)
+      end
+
+      it 'does not override the default message with a blank' do
+        expect(validating_length(type: :jsonb, is: 4)).
+          to validate_length_of(:attr).as_array.is_equal_to(4).with_message(nil)
+      end
+    end
+  end
```
Comment (mcmire) [inline spec/unit/shoulda/matchers/active_model/validate_length_of_matcher_spec.rb:555]: We should also add tests against an ActiveModel attribute that is marked as an array. There doesn't appear to be any tests that use an ActiveModel model in this file, so you may have to define a new method that constructs such a model instead of an ActiveRecord model. Instead of the `define_model` helper, you should be able to use the `define_active_model_class` helper to do so. See: https://github.com/jarenas9539/shoulda-matchers/blob/feature/extending-active-model-validating-length-of-to-allow-arrays/spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb#L1134.

**Review by mcmire — COMMENTED** [review]
These changes make sense to me. Just had a few comments.

File: lib/shoulda/matchers/active_model/validate_absence_of_matcher.rb (hunk 1 above)
Comment (jarenas9539) [reply lib/shoulda/matchers/active_model/validate_absence_of_matcher.rb:146]: I decided to do it in this PR to avoid repetition of `array_column?` and to avoid forgetting to remove it later 😅, I'll roll it back and will do the change on another PR after this one gets merged 👍

**Comment by jarenas9539** [comment]
@mcmire changes applied. This one is ready for another iteration. 👍

**Review by mcmire — COMMENTED** [review]
Looking better! I just had two more suggestions related to documentation.

File: lib/shoulda/matchers/active_model/validate_length_of_matcher.rb [hunk 4]
```diff
@@ -260,6 +260,27 @@ module ActiveModel
       #       should validate_length_of(:bio).is_at_least(15).allow_blank
       #     end
       #
+      # ##### as_array
+      #
+      # Use `as_array` to force the attribute being validate with array values
+      #
+      #     class User
+      #       include ActiveModel::Model
+      #       attr_accessor :arr, array: true
```
[the remark this answers is not in the document]
Comment (jarenas9539) [reply lib/shoulda/matchers/active_model/validate_length_of_matcher.rb]: Good catches! Already updated. Thanks 👍.

**Review by mcmire — APPROVED** [review]
Looks good! Thanks so much for the PR!
